### PR TITLE
feat: Support Parallel Segment Fetching

### PR DESCRIPTION
**Pull request by tyrelltle**
closes https://github.com/shaka-project/shaka-player/issues/4658.

This solution is inspired by abandoned PR https://github.com/shaka-project/shaka-player/pull/2809, which implements segment prefetching ahead of current play head. 

image

### CODE REVIEW

**Comment by avelad** [comment]
Thanks for your patience, I'll try to review it throughout this week.

File: externs/shaka/player.js [hunk 1]
```diff
@@ -1077,6 +1078,12 @@ shaka.extern.ManifestConfiguration;
  *   start date will not change, and would save parsing the segment multiple
  *   times needlessly.
  *   Defaults to <code>false</code>.
+ * @property {boolean} segmentPrefetchLimit
+ *   The maximum number of segments for each active stream to be prefetched
+ *   ahead of playhead in parallel.
+ *   If <code>null</code> or <code><1</code>, the segments will be fetched
+ *   sequentially.
+ *   Defaults to <code>null</code>.
```
Comment (avelad) [inline externs/shaka/player.js:1086]: Defaults to <code>0</code>.

File: externs/shaka/player.js [hunk 2]
```diff
@@ -1077,6 +1078,12 @@ shaka.extern.ManifestConfiguration;
  *   start date will not change, and would save parsing the segment multiple
  *   times needlessly.
  *   Defaults to <code>false</code>.
+ * @property {boolean} segmentPrefetchLimit
+ *   The maximum number of segments for each active stream to be prefetched
+ *   ahead of playhead in parallel.
+ *   If <code>null</code> or <code><1</code>, the segments will be fetched
```
Comment (avelad) [inline externs/shaka/player.js:1084]: null is not a valid value

File: lib/media/streaming_engine.js [hunk 3]
```diff
@@ -170,6 +171,23 @@ shaka.media.StreamingEngine = class {
...
         new shaka.net.Backoff(failureRetryParams, autoReset);
+
+    // Allow configuring the segment prefetch in middle of the playback.
+    for (const type of this.mediaStates_.keys()) {
+      const state = this.mediaStates_.get(type);
+      if (state.segmentPrefetch) {
+        state.segmentPrefetch.resetLimit(config.segmentPrefetchLimit);
+        if (!(config.segmentPrefetchLimit > 0)) {
+          /**
+           * ResetLimit is still needed in this case,
```
Comment (avelad) [inline lib/media/streaming_engine.js:182]: Please use a // comment

File: lib/media/streaming_engine.js [hunk 4]
```diff
@@ -1091,12 +1133,15 @@ shaka.media.StreamingEngine = class {
...
+    if (mediaState.segmentPrefetch && mediaState.segmentIterator) {
+      mediaState.segmentPrefetch.prefetchSegments(reference);
+    }
+
     const p = this.fetchAndAppend_(mediaState, presentationTime, reference);
     p.catch(() => {});  // TODO(#1993): Handle asynchronous errors.
     return null;
   }
 
-
```
Comment (avelad) [inline lib/media/streaming_engine.js:1099]: Please revert this change

File: lib/media/streaming_engine.js [hunk 5]
```diff
@@ -2000,6 +2045,35 @@ shaka.media.StreamingEngine = class {
...
+   * Fetches the given segment.
+   *
+   * @param {!shaka.extern.Stream} stream
+   * @param {(!shaka.media.InitSegmentReference|!shaka.media.SegmentReference)}
+   *   reference
+   * @param {?function(BufferSource):!Promise=} streamDataCallback
+   *
+   * @return {!shaka.net.NetworkingEngine.PendingRequest}
+   * @private
+   * @suppress {strictMissingProperties}
```
Comment (avelad) [inline lib/media/streaming_engine.js:2074]: Why?

**Review by avelad — CHANGES_REQUESTED** [review]

File: lib/media/streaming_engine.js (hunk 5 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js:2074]: It was originally copied from fetch_() as this method was extracted from that.
I just tried remove it but got following err. Added instanceof check and working now.


```
[JSC_STRICT_INEXISTENT_UNION_PROPERTY] Property startTime not defined on all member types of (shaka.media.InitSegmentReference|shaka.media.SegmentReference)
  2091|           duration: reference.endTime - reference.startTime,
                                                          ^^^^^^^^^
```

File: lib/media/streaming_engine.js (hunk 3 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js]: updated

File: lib/media/streaming_engine.js (hunk 4 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js]: updated

File: externs/shaka/player.js (hunk 2 above)
Comment (tyrelltle) [reply externs/shaka/player.js]: changed to 0

File: externs/shaka/player.js (hunk 1 above)
Comment (tyrelltle) [reply externs/shaka/player.js]: updated

**Comment by joeyparrish** [comment]
I skimmed this PR, and it looks interesting and well-constructed.  You've designed something that appears to address most of my concerns, it defaults to 0 (good for low-memory devices), and the prefetcher cleans up after itself so it doesn't hold onto an unbounded number of segments.  This could also be useful in a future implementation of #880 (preload API).

I haven't read it in detail, and I have some deadlines to get AV sync fixes done this week before we go on holidays at Google.  But I will come back to this in January for a detailed review, and I expect this is something that will ultimately ship in Shaka Player.

Thanks for your contribution!

**Comment by tyrelltle** [comment]
Thanks @joeyparrish !
Assuming this change goes into main branch in January, when a new version that has the change will be released ?
Asking so we can plan accordingly in case the change is not expected to be released soon.
Thanks !

**Comment by joeyparrish** [comment]
We can make a feature release at any time.  We don't have a fixed release cycle.

**Comment by avelad** [comment]
@tyrelltle can you rebase and resolve the conflicts? Thanks!

**Comment by tyrelltle** [comment]
@avelad yes I just rebased now

**Review by avelad — APPROVED** [review]

File: lib/media/segment_prefetch.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ * ahead of playhead, to reduce the chances of rebuffering.
+ */
+shaka.media.SegmentPrefetch = class {
+  /**
+   * @param {number} prefetchLimit
+   * @param {shaka.extern.Stream} stream
+   * @param {function(
+   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+   *  shaka.extern.Stream
+   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
```
Comment (theodab) [inline lib/media/segment_prefetch.js:26]: I would define the `fetchDispatcher` function with a `@typedef`, so it doesn't need to be written out each time like this.
See, for reference, `shaka.net.NetworkingEngine.OnHeadersReceived` inside `lib/net/networking_engine.js`.

File: lib/media/segment_prefetch.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+   */
+  constructor(prefetchLimit, stream, fetchDispatcher) {
+    /**
+     * @private {number}
```
Comment (theodab) [inline lib/media/segment_prefetch.js:30]: This block could be single-line.

File: lib/media/segment_prefetch.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    /** @private {function(
+     * !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+     * shaka.extern.Stream
+     * ):!shaka.net.NetworkingEngine.PendingRequest} */
+    this.fetchDispatcher_ = fetchDispatcher;
+
+    /**
+     * @private {!Map.<shaka.media.SegmentReference,
+                 !shaka.net.NetworkingEngine.PendingRequest>}
```
Comment (theodab) [inline lib/media/segment_prefetch.js:48]: This second line of the comment here should have an asterisk at the start too.

File: lib/media/segment_prefetch.js [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  /**
+   * Fetch next segments ahead of current segment.
+   *
+   * @param {(!shaka.media.SegmentReference)} startReference
+   * @public
+   */
+  prefetchSegments(startReference) {
+    if (!(this.prefetchLimit_ > 0)) {
```
Comment (theodab) [inline lib/media/segment_prefetch.js:60]: `prefetchLimit_` is non-nullable, so couldn't this just be:
```js
if (this.prefetchLimit_ <= 0) {
```

File: lib/media/segment_prefetch.js [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
Comment (theodab) [inline lib/media/segment_prefetch.js:63]: This could be compacted into one line:
```js
const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(this.stream_);
```

File: lib/media/segment_prefetch.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        this.stream_,
+    );
+    if (!this.stream_.segmentIndex) {
+      shaka.log.info(
```
Comment (theodab) [inline lib/media/segment_prefetch.js:67]: This could be compacted into one line too.

File: lib/media/segment_prefetch.js [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+   * Get the result of prefetched segment if already exists.
+   * @param {(
+   *  !shaka.media.InitSegmentReference|!shaka.media.SegmentReference
+   * )} reference
+   * @return {?shaka.net.NetworkingEngine.PendingRequest} op
+   * @public
+   */
+  getPrefetchedSegment(reference) {
+    let op = null;
+    if (
```
Comment (theodab) [inline lib/media/segment_prefetch.js:97]: I assume that returning null here if the input is not a SegmentReference is to leave open the possibility of fetching init segments at a later date?
If so, it should probably be a separate return clause, with a comment explaining that it's a placeholder.
That would also remove the need for this multi-line if-statement.

File: lib/media/segment_prefetch.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+      !(reference instanceof shaka.media.SegmentReference) ||
+      !(this.prefetchLimit_ > 0)
+    ) {
+      return null;
+    }
+
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
Comment (theodab) [inline lib/media/segment_prefetch.js:104]: This could be compacted into one line.

File: lib/media/segment_prefetch.js [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        this.stream_,
+    );
+
+    if (this.segmentPrefetchMap_.has(reference)) {
+      op = this.segmentPrefetchMap_.get(reference);
```
Comment (theodab) [inline lib/media/segment_prefetch.js:109]: I don't think you really need to define the `op` variable at the top of this function, since it's only set here anyway.
You could just declare it as a local variable in this clause, and the other clause could just return null.

File: lib/media/segment_prefetch.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    const iterator = this.stream_.segmentIndex.getIteratorForTime(maxTime);
+    let reference = startReference;
+    while (this.segmentPrefetchMap_.size < this.prefetchLimit_ &&
+            reference != null) {
+      if (!this.segmentPrefetchMap_.has(reference)) {
+        const op = this.fetchDispatcher_(reference, this.stream_);
+        this.segmentPrefetchMap_.set(reference, op);
+      }
+      this.prefetchPosTime_ = reference.startTime;
+      reference =iterator.next().value;
```
Comment (theodab) [inline lib/media/segment_prefetch.js:83]: There should be a space after the `=`.

File: lib/media/segment_prefetch.js [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+  /**
+   * Clear all segment data.
+   * @public
+   */
+  clearAll() {
+    if (this.segmentPrefetchMap_.size === 0) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
Comment (theodab) [inline lib/media/segment_prefetch.js:132]: This could be compacted into one line.

File: lib/media/segment_prefetch.js [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+   * @public
+   */
+  switchStream(stream) {
+    if (stream && stream !== this.stream_) {
+      this.clearAll();
+      this.stream_ = stream;
+    }
+  }
+
+  /** @private */
```
Comment (theodab) [inline lib/media/segment_prefetch.js:169]: This JSDoc should list the input parameter, and the return value.

File: lib/media/segment_prefetch.js [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+  /**
+   * Reset the prefetchLimit and clear all internal states.
+   * Called by StreamingEngine when configure() was called.
+   * @param {number} prefetchLimit
+   * @public
+   */
+  resetLimit(prefetchLimit) {
+    if (prefetchLimit !== this.prefetchLimit_) {
+      this.clearAll();
```
Comment (theodab) [inline lib/media/segment_prefetch.js:152]: Design-wise, I'm not sure it makes sense to always clear all of the segments away in this case.
For example, if you are going from a pre-fetch limit of 4 to a pre-fetch limit of 6, couldn't you just keep the stuff you have already pre-fetched?

It seems like the most efficient thing to do would be to store an array that keeps track of the order of the pre-fetched segments, and then pop and clear segments from the end of it until you fit into the new limit. That way nothing happens if the limit has been increased, and you clear away as few pre-fetched segments as possible in the case of the limit being decreased.

File: lib/media/streaming_engine.js [hunk 19]
```diff
@@ -818,6 +838,7 @@ shaka.media.StreamingEngine = class {
       stream,
       type: stream.type,
       segmentIterator: null,
+      segmentPrefetch: this.createSegmentPrefetch_(stream),
```
Comment (theodab) [inline lib/media/streaming_engine.js:841]: We probably should only make prefetchers for video and audio streams, not text or image.

File: lib/media/streaming_engine.js [hunk 20]
```diff
@@ -170,6 +171,21 @@ shaka.media.StreamingEngine = class {
...
+
+    // Allow configuring the segment prefetch in middle of the playback.
+    for (const type of this.mediaStates_.keys()) {
+      const state = this.mediaStates_.get(type);
+      if (state.segmentPrefetch) {
+        state.segmentPrefetch.resetLimit(config.segmentPrefetchLimit);
+        if (!(config.segmentPrefetchLimit > 0)) {
+          // ResetLimit is still needed in this case,
+          // to abort existing prefetch operations.
+          state.segmentPrefetch = null;
```
Comment (theodab) [inline lib/media/streaming_engine.js:183]: If we clear the `segmentPrefetch` object away if the limit is reduced to 0, we might not need all of those if-statements that check to see if the limit is >0 inside the class itself.
Instead, you should turn those into asserts.

File: lib/media/streaming_engine.js [hunk 21]
```diff
@@ -2027,28 +2099,25 @@ shaka.media.StreamingEngine = class {
         streamDataCallback);
 
     shaka.log.v2('fetching: reference=', reference);
-
-    const stream = mediaState.stream;
+    let duration = 0;
+    if ((reference instanceof shaka.media.SegmentReference)) {
```
Comment (theodab) [inline lib/media/streaming_engine.js:2103]: You don't need the second set of parentheses here.

File: lib/media/streaming_engine.js [hunk 22]
```diff
@@ -2336,6 +2414,8 @@ shaka.media.StreamingEngine.PlayerInterface;
  *   updating.
  * @property {shaka.net.NetworkingEngine.PendingRequest} operation
  *   Operation with the number of bytes to be downloaded.
+ * @property {shaka.media.SegmentPrefetch} segmentPrefetch
+ *   segmentPrefetch.
```
Comment (theodab) [inline lib/media/streaming_engine.js:2418]: You could use an actual description here.
Perhaps something like...

> A prefetch object for managing prefetching. Null if unneeded (if prefetching is disabled, etc).

File: lib/media/streaming_engine.js [hunk 23]
```diff
@@ -2336,6 +2414,8 @@ shaka.media.StreamingEngine.PlayerInterface;
  *   updating.
  * @property {shaka.net.NetworkingEngine.PendingRequest} operation
  *   Operation with the number of bytes to be downloaded.
+ * @property {shaka.media.SegmentPrefetch} segmentPrefetch
```
Comment (theodab) [inline lib/media/streaming_engine.js:2417]: This should probably be `@property {?shaka.media.SegmentPrefetch}`, to make it clearer that this is expected to sometimes be null.

File: test/test/util/fake_segment_prefetch.js [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+ */
+shaka.test.FakeSegmentPrefetch = class {
+  constructor(stream, segmentData) {
+    /** @private {(Set.<!shaka.media.SegmentReference>)} */
+    this.requestedReferences_ = new Set();
+
+    /** @private {shaka.extern.Stream} */
+    this.streamObj_ = stream;
+
+    /** @private {
```
Comment (theodab) [inline test/test/util/fake_segment_prefetch.js:23]: The `/**` and `*/` should be on separate lines. Also, this could be compacted some.

```js
   /**
    * @private {!Object.<string, shaka.test.FakeMediaSourceEngine.SegmentData>}
    */
```

**Review by theodab — CHANGES_REQUESTED** [review]

**Comment by tyrelltle** [comment]
@theodab thanks for the comments !
Agree all of them and will make sure address them early next week

File: demo/locales/source.json [hunk 25]
```diff
@@ -1042,5 +1042,9 @@
   "DEMO_XLINK_SEARCH": {
     "description": "A tooltip for an optional search term.",
     "message": "Filters for assets that have [JARGON:XLINK] tags in their manifests, so that they can be broken into multiple files."
+  },
+  "DEMO_SEGMENT_PREFETCH_LIMIT": {
+    "description": "Segment Prefetch Limit configuration used by StreamingEngine.",
```
[the remark this answers is not in the document]
Comment (tyrelltle) [reply demo/locales/source.json]: updated to not mention StreamingEngine

File: lib/media/segment_prefetch.js (hunk 7 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: updated

File: lib/media/streaming_engine.js (hunk 20 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js:183]: changed the "if prefetchLimit < 0" statements that are inside SegmentPrefetch, to goog.asserts.assert

File: lib/media/streaming_engine.js (hunk 22 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js]: updated

**Comment by tyrelltle** [comment]
> @theodab thanks for the comments ! Agree all of them and will make sure address them early next week

addressed some of the comments, will finish other ones by tomorrow

File: lib/media/streaming_engine.js (hunk 23 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js]: updated

File: lib/media/segment_prefetch.js (hunk 6 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: defined type

**Review by theodab — CHANGES_REQUESTED** [review]
Thanks for the work, by the way.

File: lib/media/segment_prefetch.js (hunk 18 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: updated. actually it turns out that Map.keys reserves the insertion order.
so instead of creating a new ordered array, i just grab `segmentPrefetchMap_.keys` and iterate from end to pop + clear the segments.
not sure Array.from(iterator) will actually do any iteration though, but it might have minimal performance impact since resetLimit could be called very rarely.
open to other suggestions :)

File: lib/media/segment_prefetch.js [hunk 26]
```diff
@@ -20,15 +21,10 @@ shaka.media.SegmentPrefetch = class {
   /**
    * @param {number} prefetchLimit
    * @param {shaka.extern.Stream} stream
-   * @param {function(
-   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
-   *  shaka.extern.Stream
-   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
+   * @param {shaka.media.SegmentPrefetch.fetchDispatcher} fetchDispatcher
```
[the remark this answers is not in the document]
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: sorry missed that, updated

File: lib/media/segment_prefetch.js (hunk 9 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: replaced with assert

File: lib/media/segment_prefetch.js (hunk 12 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: actually it was a mistake to leave InitSegmentReferece as an input type, so i changed the input type to only be SegmentReference

File: lib/media/streaming_engine.js (hunk 19 above)
Comment (tyrelltle) [reply lib/media/streaming_engine.js:841]: updated createSegmentPrefetch_ to return null if stream.type is not VIDEO nor AUDIO

**Comment by tyrelltle** [comment]
> Thanks for the work, by the way.

Thanks for the comments I updated all of them. Let me know if any other suggestions :)

File: lib/media/segment_prefetch.js (hunk 18 above)
Comment (tyrelltle) [reply lib/media/segment_prefetch.js]: pushed a new commit to simplify resetLimit, by using while + arr.pop() instead of for loop from end of arr

**Review by avelad — APPROVED** [review]

**Review by theodab — APPROVED** [review]
